janitor/functions.py

```diff
@@ -3140,25 +3094,67 @@ def update_where(
         # 4 8  0
 
     :param df: The pandas DataFrame object.
-    :param conditions: conditions used to update a target column and target
-        value
+    :param conditions: Conditions used to update a target column
+                       and target value.
     :param target_column_name: Column to be updated. If column does not exist
-        in dataframe, a new column will be created; note that entries that do
-        not get set in the new column will be null.
+                               in the dataframe, a new column will be created;
+                               note that entries that do not get set
+                               in the new column will be null.
     :param target_val: Value to be updated
     :returns: An updated pandas DataFrame.
     :raises: IndexError if **conditions** does not have the same length as
-        **df**.
+             **df**.
+    :raises: TypeError if **conditions** is not a pandas-compatible string
+             query.
     """
-    df.loc[conditions, target_column_name] = target_val
+
+    # Boolean expressions will be deprecated in future versions
+    # query style method used instead
+    # to accommodate changes if the dataframe has been modified
+    # somewhere along the method chain block
+    # query style method suggested by @zbarry
+    # github issue #663
+
+    # this section tests for boolean expressions
+    # and includes a deprecation warning to the user
+    if isinstance(conditions,
```

Comment:
I would flip the order of the type checking happening here. First, check if `conditions` is a string. If it is, take the `.query` path. If it's not, just go ahead and plug it right into `df.loc[conditions, target_column_name] = target_val` without the check that it is a `pd.Series`, and especially one that contains booleans. There are a bunch of valid types you can use to index a `DataFrame`'s rows, and we shouldn't be the ones gating that functionality from users.